pkg/kubelet/kubelet.go

```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
 	kl.syncLoop(updates, kl)
 }
 
+func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{Name: kl.hostname},
+		Status: api.NodeStatus{
+			Capacity: api.ResourceList{
+				api.ResourceCPU:    *resource.NewMilliQuantity(1000, resource.DecimalSI),
+				api.ResourceMemory: resource.MustParse("3Gi"),
+			},
+		},
+	}
+	if kl.cloud != nil {
+		instances, ok := kl.cloud.Instances()
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Inspect the local machine rather than calling out
+		// to the cloud provider.
+		resources, err := instances.GetNodeResources(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node resources from cloud provider: %v", err)
+		}
+		if resources != nil {
+			node.Status.Capacity = resources.Capacity
+		}
+		instanceID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node address from cloud provider: %v", err)
+		}
+		node.Status.Addresses = nodeAddresses
+	}
+	return node, nil
+}
+
+// registerNode registers the node with the cluster master.
+func (kl *Kubelet) registerNode() error {
+	step := 100 * time.Millisecond
+	for start := time.Now(); time.Since(start) < 30*time.Minute; time.Sleep(step) {
```

Comment:
Maybe with a back off?  100ms early, every few seconds later?

----- Original Message -----
> > +			return nil, fmt.Errorf("failed to get instance ID from cloud provider:
> > %v", err)
> > +		}
> > +		node.Spec.ExternalID = instanceID
> > +		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
> > +		if err != nil {
> > +			return nil, fmt.Errorf("failed to get node address from cloud provider:
> > %v", err)
> > +		}
> > +		node.Status.Addresses = nodeAddresses
> > +	}
> > +	return node, nil
> > +}
> > +
> > +// registerNode registers the node with the cluster master.
> > +func (kl *Kubelet) registerNode() error {
> > +	step := 100 * time.Millisecond
> > +	for start := time.Now(); time.Since(start) < 30*time.Minute;
> > time.Sleep(step) {
> 
> Do you think it should just keep trying indefinitely?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r28936676